File: bindings/kotlin/src/main/kotlin/strling/core/Compiler.kt
package strling.core

import kotlinx.serialization.json.JsonPrimitive

object Compiler {
    fun compile(node: Node): IROp {
        return when (node) {
            is Alternation -> IRAlt(node.alternatives.map { compile(it) })
            is Sequence -> {
                val compiledParts = node.parts.map { compile(it) }
                val mergedParts = mutableListOf<IROp>()
                for (part in compiledParts) {
                    if (mergedParts.isNotEmpty() && mergedParts.last() is IRLit && part is IRLit) {
                        val last = mergedParts.removeAt(mergedParts.lastIndex) as IRLit
                        mergedParts.add(IRLit(last.value + part.value))
                    } else {
                        mergedParts.add(part)
                    }
                }
                if (mergedParts.size == 1) {
                    mergedParts[0]
                } else {
                    IRSeq(mergedParts)
                }
            }
            is Literal -> IRLit(node.value)
            is Dot -> IRDot()
            is Anchor -> {
                val at = if (node.at == "NonWordBoundary") "NotWordBoundary" else node.at
                IRAnchor(at)
            }
            is CharacterClass -> IRCharClass(node.negated, node.members.map { compileClassItem(it) })
            is Quantifier -> {
                val mode = when {
                    node.possessive -> "Possessive"
                    node.lazy -> "Lazy"
                    else -> "Greedy"
                }
                // If max is null, it means infinity ("Inf")
                val maxVal = node.max ?: JsonPrimitive("Inf")
                IRQuant(compile(node.target), node.min, maxVal, mode)
            }
            is Group -> {
                val atomic = if (node.atomic == true) true else null
                IRGroup(node.capturing, compile(node.body), node.name, atomic)
            }
            is Backreference -> IRBackref(node.index, node.name)
            is Lookahead -> IRLook("Ahead", false, compile(node.body))
            is NegativeLookahead -> IRLook("Ahead", true, compile(node.body))
            is Lookbehind -> IRLook("Behind", false, compile(node.body))
            is NegativeLookbehind -> IRLook("Behind", true, compile(node.body))
        }
    }

    private fun compileClassItem(item: ClassItem): IRClassItem {
        return when (item) {
            is Range -> IRClassRange(item.from, item.to)
            is Escape -> {
                val type = when (item.kind) {
                    "digit" -> "d"
                    "space" -> "s"
                    "word" -> "w"
                    "not-digit" -> "D"
                    "not-space" -> "S"
                    "not-word" -> "W"
                    else -> item.kind
                }
                IRClassEscape(type)
            }
            is Literal -> IRClassLiteral(item.value)
            is UnicodeProperty -> {
                val type = if (item.negated) "P" else "p"
                IRClassEscape(type, item.value)
            }
        }
    }
}

File: bindings/csharp/src/STRling/Core/Compiler.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Strling.Core;

public class Compiler
{
    public static IROp Compile(Node node)
    {
        return node switch
        {
            Alt alt => new IRAlt((alt.Alternatives ?? new List<Node>()).Select(Compile).ToList()),
            Seq seq => CreateSeq((seq.Parts ?? new List<Node>()).Select(Compile)),
            Lit lit => new IRLit(lit.Value),
            Dot dot => new IRDot(),
            Anchor anchor => new IRAnchor(MapAnchor(anchor.At)),
            CharClass cc => new IRCharClass(cc.Negated, (cc.Members ?? new List<ClassItem>()).Select(CompileClassItem).ToList()),
            Quant quant => CompileQuant(quant),
            Group group => new IRGroup(group.Capturing, Compile(group.Body), group.Name, group.Atomic),
            Backref backref => new IRBackref(backref.Index, backref.Name),
            Lookahead la => new IRLook("Ahead", false, Compile(la.Body)),
            NegativeLookahead nla => new IRLook("Ahead", true, Compile(nla.Body)),
            Lookbehind lb => new IRLook("Behind", false, Compile(lb.Body)),
            NegativeLookbehind nlb => new IRLook("Behind", true, Compile(nlb.Body)),
            _ => throw new NotImplementedException($"Unknown node type: {node.GetType().Name}")
        };
    }

    private static IROp CreateSeq(IEnumerable<IROp> parts)
    {
        var list = new List<IROp>();
        foreach (var part in parts)
        {
            if (part is IRSeq subSeq)
            {
                list.AddRange(subSeq.Parts);
            }
            else
            {
                list.Add(part);
            }
        }

        var merged = new List<IROp>();
        foreach (var op in list)
        {
            if (merged.Count > 0 && merged[^1] is IRLit lastLit && op is IRLit newLit)
            {
                merged[^1] = new IRLit(lastLit.Value + newLit.Value);
            }
            else
            {
                merged.Add(op);
            }
        }

        if (merged.Count == 1) return merged[0];
        return new IRSeq(merged);
    }

    private static string MapAnchor(string at)
    {
        return at == "NonWordBoundary" ? "NotWordBoundary" : at;
    }

    private static IRClassItem CompileClassItem(ClassItem item)
    {
        return item switch
        {
            ClassRange range => new IRClassRange(range.From, range.To),
            ClassLiteral lit => new IRClassLiteral(lit.Value),
            ClassEscape esc => new IRClassEscape(MapEscapeKind(esc.Kind), null),
            ClassUnicodeProperty prop => new IRClassEscape(prop.Negated ? "P" : "p", prop.Value),
            _ => throw new NotImplementedException($"Unknown class item type: {item.GetType().Name}")
        };
    }

    private static string MapEscapeKind(string kind)
    {
        return kind switch
        {
            "digit" => "d",
            "not-digit" => "D",
            "word" => "w",
            "not-word" => "W",
            "whitespace" => "s",
            "not-whitespace" => "S",
            "space" => "s",
            "not-space" => "S",
            _ => kind
        };
    }

    private static IROp CompileQuant(Quant quant)
    {
        string mode = "Greedy";
        if (quant.Lazy) mode = "Lazy";
        else if (quant.Possessive) mode = "Possessive";

        object max = quant.Max.HasValue ? quant.Max.Value : "Inf";

        return new IRQuant(Compile(quant.Target), quant.Min, max, mode);
    }
}
